src/rules/noObjectComparisonRule.ts

```diff
@@ -0,0 +1,181 @@
+        ruleName: "no-object-comparison",
+        description: "Only allow comparisons between primitives.",
+        optionsDescription: Lint.Utils.dedent`
+                One argument may be optionally provided:
+                * \`${OPTION_ALLOW_EQUAL}\` allows \`!=\` \`==\` \`!==\` \`===\` comparison between any types.`,
+        options: {
+            type: "array",
+            items: {
+                type: "string",
+                enum: [OPTION_ALLOW_EQUAL],
+            },
+            minLength: 0,
+            maxLength: 1,
+        },
+        optionExamples: [true, [true, OPTION_ALLOW_EQUAL]],
+        rationale: Lint.Utils.dedent`
+                When using comparison operators to compare objects, they compare references and not values.
+                This is often done accidentally.
+                With this rule, \`>\`, \`>=\`, \`<\`, \`<=\` operators are only allowed when comparing \`numbers\`.
+                \`===\`, \`!==\` are allowed for \`number\` \`string\` and \`boolean\` types and if one of the
+                operands is \`null\` or \`undefined\`.
+            `,
+        type: "functionality",
+        typescriptOnly: false,
+        requiresTypeInfo: true,
+        codeExamples,
+    };
+    /* tslint:enable:object-literal-sort-keys */
+
+    public static INVALID_COMPARISON = `Invalid comparison`;
+
+    public applyWithProgram(sourceFile: ts.SourceFile, program: ts.Program): Lint.RuleFailure[] {
+        return this.applyWithFunction(
+            sourceFile,
+            walk,
+            {
+                allowEquals: this.ruleArguments.indexOf(OPTION_ALLOW_EQUAL) !== -1,
+            },
+            program,
+        );
+    }
+}
+
+function walk(ctx: Lint.WalkContext<Options>, program: ts.Program) {
+    const checker = program.getTypeChecker();
+
+    return ts.forEachChild(ctx.sourceFile, function cb(node: ts.Node): void {
+        if (
+            isBinaryExpression(node) &&
+            isComparisonOperator(node) &&
+            !(isAnyType(node.left, checker) || isAnyType(node.right, checker)) &&
+            !(isNumericType(node.left, checker) || isNumericType(node.right, checker)) &&
+            !(
+                isEqualityOperator(node) &&
+                (isStringType(node.left, checker) || isStringType(node.right, checker))
+            ) &&
+            !(
+                isEqualityOperator(node) &&
+                (isBooleanType(node.left, checker) || isBooleanType(node.right, checker))
+            ) &&
+            !(
+                isEqualityOperator(node) &&
+                (isEnumType(node.left, checker) || isEnumType(node.right, checker))
+            ) &&
+            !(
+                isEqualityOperator(node) &&
+                (isNullOrUndefinedType(node.left, checker) ||
+                    isNullOrUndefinedType(node.right, checker))
+            ) &&
+            !(ctx.options.allowEquals && isEqualityOperator(node))
```

Comment:
This >20 line comparison, although impressively long (would this be a record in the project?), is very difficult to read and seems to be doing a lot of checks multiple time. `isEqualityOperator(node)` is called a lot and many of the utility functions call `checker.getTypeAtLocation`.

Please split this logic out into a new function with separate conditionals so you can save the results of those checks & avoid some of the duplicate work.

Reply:
Yeah, you are right, not the most readable statement. 😅